### PR TITLE
bpf: make config as a constant

### PR DESCRIPTION
**Pull request by Asphaltt**
Since 5.2 kernel, [bpf: implement lookup-free direct value access for maps](https://lore.kernel.org/bpf/20190409210910.32048-2-daniel@iogearbox.net/).

Then, use `bpf2bpf` to reduce .o file size.

Without `bpf2bpf`, the size of .o files:

```bash
-rwxrwx--- 1 root vboxsf 122K Feb  4 13:33 kprobemultipwru_bpfeb.o
-rwxrwx--- 1 root vboxsf 121K Feb  4 13:33 kprobemultipwru_bpfel.o
-rwxrwx--- 1 root vboxsf 117K Feb  4 13:34 kprobemultipwruwithoutoutputskb_bpfeb.o
-rwxrwx--- 1 root vboxsf 116K Feb  4 13:34 kprobemultipwruwithoutoutputskb_bpfel.o
-rwxrwx--- 1 root vboxsf 122K Feb  4 13:33 kprobepwru_bpfeb.o
-rwxrwx--- 1 root vboxsf 121K Feb  4 13:33 kprobepwru_bpfel.o
-rwxrwx--- 1 root vboxsf 117K Feb  4 13:34 kprobepwruwithoutoutputskb_bpfeb.o
-rwxrwx--- 1 root vboxsf 117K Feb  4 13:34 kprobepwruwithoutoutputskb_bpfel.o
```

With `bpf2bpf`, the size of .o files:

```bash
-rwxrwx--- 1 root vboxsf 67K Feb  4 14:05 kprobemultipwru_bpfeb.o
-rwxrwx--- 1 root vboxsf 67K Feb  4 14:05 kprobemultipwru_bpfel.o
-rwxrwx--- 1 root vboxsf 66K Feb  4 14:05 kprobemultipwruwithoutoutputskb_bpfeb.o
-rwxrwx--- 1 root vboxsf 66K Feb  4 14:05 kprobemultipwruwithoutoutputskb_bpfel.o
-rwxrwx--- 1 root vboxsf 67K Feb  4 14:05 kprobepwru_bpfeb.o
-rwxrwx--- 1 root vboxsf 67K Feb  4 14:05 kprobepwru_bpfel.o
-rwxrwx--- 1 root vboxsf 66K Feb  4 14:05 kprobepwruwithoutoutputskb_bpfeb.o
-rwxrwx--- 1 root vboxsf 66K Feb  4 14:05 kprobepwruwithoutoutputskb_bpfel.o
```

However, the cost is adding a bpf function call in runtime for every kprobe.

Without `bpf2bpf`, output of `bpftool prog dump xlated id ${PROGID}`:

```
int kprobe_skb_5(struct pt_regs * ctx):
; PWRU_ADD_KPROBE(5)
 0: (bf) r6 = r1
; PWRU_ADD_KPROBE(5)
 1: (79) r9 = *(u64 *)(r6 +72)
 2: (b7) r1 = 0
; struct event_t event = {};
; ...
```

With `bpf2bpf`, output of `bpftool prog dump xlated id ${PROGID}`:

```
int kprobe_skb_5(struct pt_regs * ctx):
; PWRU_ADD_KPROBE(5)
 0: (bf) r2 = r1
; PWRU_ADD_KPROBE(5)
 1: (79) r1 = *(u64 *)(r2 +72)
 2: (85) call pc+2#bpf_prog_6091cf2489a85a37_handle_everything
 3: (b7) r0 = 0
 4: (95) exit
int handle_everything(struct sk_buff * skb, struct pt_regs * ctx, bool has_get_func_ip):
; handle_everything(struct sk_buff *skb, struct pt_regs *ctx, bool has_get_func_ip) {
 5: (bf) r6 = r2
 6: (bf) r7 = r1
 7: (b7) r1 = 0
; struct event_t event = {};
; ...
```

### CODE REVIEW

File: bpf/kprobe_pwru.c [hunk 1]
```diff
@@ -84,9 +84,12 @@ struct config {
 	u8 output_tuple;
 	u8 output_skb;
 	u8 output_stack;
-	u8 pad;
+	u8 setted;
```
Comment (brb) [inline bpf/kprobe_pwru.c:87]: Nit: s/setted/is_set/

File: internal/pwru/config.go [hunk 2]
```diff
@@ -22,31 +20,32 @@ type FilterCfg struct {
...
-	//TODO: if there are more options later, then you can consider using a bit map
+	// TODO: if there are more options later, then you can consider using a bit map
 	OutputRelativeTS uint8
 	OutputMeta       uint8
 	OutputTuple      uint8
 	OutputSkb        uint8
 	OutputStack      uint8
 
-	Pad byte
+	Setted byte
```
Comment (brb) [inline internal/pwru/config.go:41]: Ditto.

**Review by brb — COMMENTED** [review]
Thanks. Nice changes!

Regarding the second patch (bpf2bpf), I'd wait until https://github.com/cilium/pwru/issues/85 has been implemented. In any case, I'm curious to know if you have measured any performance degradation when using the bpf2bpf calls?

**Comment by Asphaltt** [comment]
>  In any case, I'm curious to know if you have measured any performance degradation when using the bpf2bpf calls?

Nope.

In theory, the actual cost of `bpf2bpf` is a general function call in kernel. The cost seems like nearly 15 instructions on x86 with nearly 5 memory access ( `push` and `pop` of function call) for every `kprobe`.

😂 The first patch reduces a function call `bpf_map_lookup_elem()`. But the second patch supplements the _reduced_ function call with `bpf2bpf`. So that, there is no performance change for `pwru`. But the size of `.o` files reduces in half nearly.

**Review by brb — APPROVED** [review]
Thanks!

**Pull request merged**